proposals/NNNN-constant-buffers.md

```diff
@@ -0,0 +1,132 @@
+
+> Note: The most tricky part of the layout encoding are probably arrays of structures because the structure-specific layout gets repeated many times and might not be easy to compress. One idea how to solve this could be translating structures embedded in `cbuffer` into separate target types with their own encoded layout and including them in the `cbuffer` target type. It is not clear though if this is possible (probably yes) or if it would actually make things easier or not.
+
+Another way could be introducing a typedef concept into the LLVM IR textual representation so that the full LLVM target type with long layout representation could occur just once.
+
+### Lowering ConstantBuffer to LLVM IR
+
+The result of codegen for `cbuffer` and `ConstantBuffer` code should be identical.
+
+### Lowering Constant Buffer Variable Access
+
+Accesses to `cbuffer` variables will be lowered to LLVM IR as memory accesses in specific "resource address space" using the standard C++ structure layout rules.
+
+### DXIL Lowering
+
+LLVM pass `DXILResourceAccess` will translate these specific "resource address space" memory accesses into cbuffer DXIL ops adjusting the offsets using the cbuffer layout information encoded in `target("dx.CBuffer", ..)`. That means translating standand C++ structure layout offsets to cbuffer layout offsets and replacing the memory accesses with `llvm.dx.cbufferBufferLoad`, `llvm.dx.cbufferBufferStore`, and `extractelement ` instructions. The load and store instructions will be later lowered to `cbufferLoadLegacy` and `cbufferStoreLegacy` DXIL ops.
+
+### Handle initialization
+
+Constant buffers will be initialized the same way as other resources using the `createHandleFromBinding` intrinsics. Module initialization code need to be updated to include all constant buffers declared in a shader.
+
+## Detailed design
+
+*TBD*
+
+## Alternatives considered
+
+- Generate access to `cbuffer` varibles as memory accesses with the offset based on cbuffer layout and treat cbuffers as one big type-less memory blob in LLVM IR.
+  - There is a concern that losing the type information could lead to unnecessary copying of values.
+
+- Using type annotations to store cbuffer layout information. Subtypes would have its own layout annotations.
+  - This is something we can fall back to if encoding the cbuffer layout on LLVM target type turns out to be too unwieldy, especially when it comes to encoding the layout of an array of structures.
+
+## Open issues
+- How to encode the cbuffer layout into LLVM target type
+- How to implement `ConstantBuffer` member access
+- Handling of `$Globals` constant buffer
+- Nested `cbuffer` declarations
```

Comment:
I put together a bit of an example that illustrates how cbuffer works in various interesting cases here:
https://godbolt.org/z/E4cdx7xj8

Same shader with FXC:
https://shader-playground.timjones.io/39c0683735c0cd79b5970b5116765940

Reply:
Thanks Tex! I've created an issue to track this: https://github.com/llvm/llvm-project/issues/118524